stan/math/prim/mat/prob/normal_id_glm_lpdf.hpp

```diff
@@ -92,56 +98,149 @@ return_type_t<T_y, T_x, T_alpha, T_beta, T_scale> normal_id_glm_lpdf(
   const auto &y_val_vec = as_column_vector_or_scalar(y_val);
 
   T_scale_val inv_sigma = 1 / as_array_or_scalar(sigma_val_vec);
+  Matrix<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma_mat(N);
+  auto y_minus_mu_over_sigma = y_minus_mu_over_sigma_mat.array();
 
-  Array<T_partials_return, Dynamic, 1> y_minus_mu_over_sigma
-      = x_val * beta_val_vec;
-  y_minus_mu_over_sigma = (as_array_or_scalar(y_val_vec) - y_minus_mu_over_sigma
-                           - as_array_or_scalar(alpha_val_vec))
-                          * inv_sigma;
-
-  // Compute the derivatives.
   operands_and_partials<T_y, T_x, T_alpha, T_beta, T_scale> ops_partials(
       y, x, alpha, beta, sigma);
+
   double y_minus_mu_over_sigma_squared_sum;  // the most efficient way to
-                                             // calculate this depends on
-                                             // template parameters
-  if (!is_constant_all<T_y, T_x, T_beta, T_alpha>::value) {
-    Matrix<T_partials_return, Dynamic, 1> mu_derivative
-        = inv_sigma * y_minus_mu_over_sigma;
-    if (!is_constant_all<T_y>::value) {
-      ops_partials.edge1_.partials_ = -mu_derivative;
+  // calculate this depends on
+  // template parameters
+
+#ifdef STAN_OPENCL
+  bool use_opencl = N * M * opencl_context.tuning_opts().normal_id_glm_coeff1
+                        + N * opencl_context.tuning_opts().normal_id_glm_coeff2
+                    > 1;
```

Comment:
(unrelated to this PR) I'm starting to think we should get rid of the tuning params. Do we have anyone / anytime to make a tuner? If not then I'd be for a separate PR ripping it out

Reply:
@rok-cesnovar is already working on a autotuning.

Reply:
@rok-cesnovar is already working on autotuning.

Reply:
Yes, autotuning is a pet project of mine that should see the light of day someday along with my PhD. But that damn thing is not willing to converge :)

The way we use "tuning opts" right now is actually just as a way of selecting whether to execute on the CPU or the GPU. We have circled around this topic at Stancon  the general thought is that we dont really need to be so smart about selecting the CPU or GPU. 
So I agree with you that we should throw these out and if a user specifies `STAN_OPENCL` all OpenCL supported functions will get executed on the GPU. 

This helps us with: 
- support (there were already a few users that set `STAN_OPENCL` but saw no GPU activity and that was confusing to most)
- we dont need to search for the CPU/GPU breakpoint on OpenCL PRs and test with different hardware to find an arbitrary limit that we have no way of evaluating
- cleaner code when we include OpenCL in /prim and /rev

Its the same with MPI and threading. If the overhead is not beaten by parallelization for the input size you will not see a speedup and that is just the way it is. And those that use it (should) expect it.

The autotuning I am working on will adapt the `OpenCL` implementation parameters to tune it to the device we are executing on. MCMC applications are perfect for this with the 1000+ of iterations of the same code. I was also working on a system that would help with this autoselect but I have not found a nice solution for it and bailed on that after Stancon due to the stuff mentioned above.

Reply:
So I would get rid of the "tuning opts" that are just use for selecting when to switch from CPU to GPU.